### PR TITLE
Remove global seeding and add test fixture for seed

### PR DESCRIPTION
**Pull request by lreeb**
With our recent issues with flaky tests and global seeding, I propose to use a pytest fixture to create a random (but fixed) seed for each of the reference tests.
Benefits:
* Removes all global seeding
* Avoids runing tests with always the same fixed seeds -> will spot problems earlier
* If a test fails, the seed will show up in the log and the test can be reproduced easily

What do you think?

Signed-off-by: Reeb Lucia <Lucia.Reeb@de.bosch.com>

### CODE REVIEW

File: tests/optimizers/conftest.py [hunk 1]
```diff
@@ -13,6 +14,11 @@
 objective_name = "objective"
 
 
+@pytest.fixture
+def seed():
+    return random.getrandbits(32)
```
Comment (dynobo) [inline tests/optimizers/conftest.py:19]: I'm curious: What was your reason for using `random.getrandbits(32)` over `random.randint()`? (I didn't know `getrandbits` before and had to look it up)

**Review by dynobo — APPROVED** [review]
Thanks Lucia, removing the global seed makes so much sense!
I would have chosen even a fixed seed for the unit tests, but I'm also super fine with trying a random one. If this should turn out to be too fragile and blocks PRs too often, we now easily can switch to a fixed seed :-)

**Comment by dynobo** [comment]
If you have some time, you could also do this for [blackboxopt](https://github.com/boschresearch/blackboxopt/blob/main/blackboxopt/optimizers/testing.py). Otherwise just leave me a note, I'd also happily do it.

File: tests/optimizers/conftest.py (hunk 1 above)
Comment (sfalkner) [reply tests/optimizers/conftest.py:19]: I guess the problem is that python integers are of arbitrary size and `random.randint()` needs to arguments (min and max) value. This way, we get an unsigned int with the usual 32bit size, which should work for all packages we use, i.e. numpy and torch.

**Review by sfalkner — APPROVED** [review]
Thanks @lreeb for this addition. Looks like we should use this in other places, too :)

**Comment by lreeb** [comment]
> Thanks Lucia, removing the global seed makes so much sense! I would have chosen even a fixed seed for the unit tests, but I'm also super fine with trying a random one. If this should turn out to be too fragile and blocks PRs too often, we now easily can switch to a fixed seed :-)

I think this is somewhat of a fundamental decision that we have to make - using a fixed seed would make all of our tests deterministic, thus we wouldn't have to check for flaky tests any more, and have more stability. Using a random seed will probably take some more effort to make all tests stable - but will possibly detect more errors. I don't have a strong opinion on which way to go, just didn't like the in-between status right now!

**Comment by LGro** [comment]
Given the concern that we need to test stochastic things, it seems like a very good idea to avoid fixed seeds, but to use random seeds so that we get the stochasticity as well as reproducibility. Regarding flaky tests, it's a good idea for us to retain a close look at how flaky specific optimizers are because we might need to do something about it at a certain threshold; and with random seeds that threshold becomes explicit instead of hiding behind a fixed seed that works.
Might it be worth recording our reasoning behind the seeding strategy in an ADR?

**Comment by dynobo** [comment]
> Given the concern that we need to test stochastic things, it seems like a very good idea to avoid fixed seeds, but to use random seeds so that we get the stochasticity as well as reproducibility. Regarding flaky tests, it's a good idea for us to retain a close look at how flaky specific optimizers are because we might need to do something about it at a certain threshold; 

From my experience, deterministic unit/integration tests are very important for a fluent development workflow and to identify problems fast. I'd rather _not_ want to ask myself "Failed the test because of my change or because it is flaky?".

I agree, that testing stochasticity is important as well, but I see that as a different purpose and would rather see such potential flaky tests it in an extra, separated test-set. Especially, if we see, how bad we currently are in handling failures in those (see Teams Bots channel).

**Comment by lreeb** [comment]
Good point about adding an ADR! 
@dynobo While I fully agree on the point about deterministic tests, that would also imply that we would need to seed almost every single test, since most of them use stochasticity in some way.
I'd vote on trying with random seeds for now, and if it turns out that this is not feasible or too much overhead we might revert to a fixed seed and split up the test suite in a deterministic and flaky part.

**Pull request merged**